server/tools.go

```diff
@@ -0,0 +1,414 @@
+	}
+}
+
+// TODO: simplify if possible
+type ToolParser struct {
+	tmpl        *gotmpl.Template
+	state       State
+	sb          *strings.Builder
+	toolPrefix  string
+	toolIndex   int
+	ParserState ExternalState
+	Done        bool
+}
+
+// ? move to a separate file
+// parseJSONToolCalls attempts to parse a JSON string into a slice of ToolCalls.
+// Returns parsed tool calls, a boolean indicating if the JSON is incomplete, and a boolean indicating if the tool calls were found
+func (p *ToolParser) parseJSONToolCalls(s string) ([]api.ToolCall, bool, bool) {
+	fmt.Printf("attempting to parse JSON tool calls: input=%s\n", s)
+
+	var b bytes.Buffer
+	if err := p.tmpl.Execute(&b, map[string][]api.ToolCall{
+		"ToolCalls": {
+			{
+				Function: api.ToolCallFunction{
+					Name: "@@name@@",
+					Arguments: api.ToolCallFunctionArguments{
+						"@@argument@@": 1,
+					},
+				},
+			},
+		},
+	}); err != nil {
+		fmt.Printf("failed to execute template: error=%v\n", err)
+		return nil, false, false
+	}
+
+	// this can be either a map or an array
+	var temp any
+	err := jsonv2.Unmarshal(b.Bytes(), &temp)
+	if err != nil {
+		fmt.Printf("failed to unmarshal template: error=%v\n", err)
+		return nil, false, false
+	}
+
+	var collect func(any) []map[string]any
+	collect = func(obj any) (all []map[string]any) {
+		switch o := obj.(type) {
+		case map[string]any:
+			all = append(all, o)
+			for _, v := range o {
+				all = append(all, collect(v)...)
+			}
+		case []any:
+			for _, v := range o {
+				all = append(all, collect(v)...)
+			}
+		default:
+			// TODO: err or fallback
+			fmt.Printf("collect encountered unknown type: type=%T\n", obj)
+			return nil
+		}
+
+		return all
+	}
+
+	var templateObjects []map[string]any
+	switch t := temp.(type) {
+	case map[string]any:
+		templateObjects = []map[string]any{t}
+	case []map[string]any:
+		templateObjects = t
+	// ! fallback?
+	case []any:
+		templateObjects = collect(t)
+	}
+	if len(templateObjects) == 0 {
+		fmt.Println("no template objects found")
+		return nil, false, false
+	}
+
+	// find the keys that correspond to the name and arguments fields
+	var name, arguments string
+	for k, v := range templateObjects[0] {
+		switch v.(type) {
+		case string:
+			name = k
+			fmt.Printf("found name field: key=%s\n", k)
+		case map[string]any:
+			arguments = k
+			fmt.Printf("found arguments field: key=%s\n", k)
+		}
+	}
+
+	if name == "" || arguments == "" {
+		fmt.Printf("missing required fields: name_found=%v arguments_found=%v\n", name != "", arguments != "")
+		return nil, false, false
+	}
+
+	// TODO: there is probably some underlying repeat work here to avoid
+	// This incrementally decodes the JSON string and returns the first parsedobject
+	dec := jsontext.NewDecoder(strings.NewReader(s))
+	if got, err := dec.ReadValue(); err == nil {
+		s = got.String()
+		fmt.Printf("decoded JSON value: value=%s\n", s)
+	}
+
+	var responseObjects any
+	err = jsonv2.Unmarshal([]byte(s), &responseObjects)
```

Comment:
Why is this part of the parser working with an array of objects?

If I understand the logic right we should be able to parse one json object at a time, retrieve the tool call from that, and emit it in one `for{` of the `parser.Add` (what you have now as `toolParser.ParseToolCalls`). That way even if an array of tools is provided, we can parse it incrementally just like any other response containing multiple json objects.

Reply:
Yes but we're not stripping `[` as part of the prefix right now. Once I have that in (currently as a follow-up), we'll be operating on the inside of the list.